azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,122 @@
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = PriorityQueue()
+
+    @staticmethod
+    def get_thread_name():
+        return TelemetryServiceHandler._THREAD_NAME
+
+    def run(self):
+        logger.info("Start Extension Telemetry service.")
+        self.start()
+
+    def is_alive(self):
+        return self._thread is not None and self._thread.is_alive()
+
+    def start(self):
+        self._thread = threading.Thread(target=self._process_telemetry_thread)
+        self._thread.setDaemon(True)
+        self._thread.setName(self.get_thread_name())
+        self._thread.start()
+
+    def stop(self):
+        """
+        Stop server communication and join the thread to main thread.
+        """
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def join(self):
+        self._queue.join()
+        self._thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def enqueue_event_func(self, event):
+        # Add event to queue and set event
+        self._queue.put(event)
+
+        # Set the event if any enqueue happens (even if already set) to trigger sending those events
+        self._should_process_events.set()
+
+    def _process_telemetry_thread(self):
+        logger.info("Successfully started the {0} thread".format(self.get_thread_name()))
+        try:
+            # On demand wait, start processing as soon as there is any data available in the queue
+            # In worst case, also keep checking every 5 mins to ensure that no data is being missed
+            while not self.stopped():
+                self._should_process_events.wait(timeout=TelemetryServiceHandler._MAX_TIMEOUT)
+                self._send_events_in_queue()
+
+        except Exception as error:
+            err_msg = "An unknown error occurred in the {0} thread main loop, stopping thread. Error: {1}, Stack: {2}".format(
+                self.get_thread_name(), ustr(error), traceback.format_exc())
+            add_event(op=WALAEventOperation.UnhandledError, message=err_msg, is_success=False)
+
+    def _get_events_in_queue(self):
+        while not self._queue.empty():
+            try:
+                event = self._queue.get()
+                yield event
+            except Exception as error:
+                logger.error("Some exception when fetching event from queue: {0}, {1}".format(ustr(error),
+                                                                                              traceback.format_exc()))
+            finally:
+                self._queue.task_done()
+
+    def _send_events_in_queue(self):
+        # Process everything in Queue
+        if not self._queue.empty():
+            self._protocol.report_event(self._get_events_in_queue)
+
+        # Reset the event when done processing all events in queue
+        if self._should_process_events.is_set() and self._queue.empty():
```

Comment:
> and self._queue.empty()

Introduces the possibility of the telemetry thread keep sending events continuously. Previously add_event would cap the number of events to 1,000 and the monitor thread would periodically take whatever number of events were available at some point in time, send them and wait for 1 minute (or was it 5?). During that minute add_event would continue capping the number of events. Seems like now the telemetry thread can keep sending events all the time.

Reply:
>Introduces the possibility of the telemetry thread keep sending events continuously.

But isint that a good thing? That is what this new thread is there for. `add_event` still caps events at 1000, I didnt change that logic. The major things that I did change were to - 
- Enqueue event as soon as we parse the file and create an object vs parsing all files and then sending out all events at once. The benefit this gives us is that now we dont have to wait to parse all files 
- Earlier since the same thread was doing parsing and sending, we would block the thread until the operation was complete and then wait another min before repeating the cycle. Now we parse the event files every minute and the telemetry thread keeps on sending it out always, ensuring that we capture more events and not miss out on anything. 

We can discuss more about this offline if you have more questions/doubts

Reply:
The batching addresses my concern. Thanks for adding it.